presto-main/src/main/java/com/facebook/presto/operator/SpillableGroupedTopNBuilder.java

```diff
@@ -0,0 +1,308 @@
+
+    private final long unspillMemoryLimit;
+
+    private Optional<Spiller> spiller = Optional.empty();
+    private ListenableFuture<?> spillInProgress = immediateFuture(null);
+    private Optional<MergeHashSort> mergeHashSort = Optional.empty();
+    private boolean producingOutput;
+
+    public SpillableGroupedTopNBuilder(
+            List<Type> sourceTypes,
+            List<Type> partitionTypes,
+            List<Integer> partitionChannels,
+            Supplier<InMemoryGroupedTopNBuilder> groupedTopNBuilderSupplier,
+            long unspillMemoryLimit,
+            OperatorContext operatorContext,
+            SpillerFactory spillerFactory)
+    {
+        this.groupedTopNBuilderSupplier = groupedTopNBuilderSupplier;
+        this.spillerFactory = spillerFactory;
+        this.sourceTypes = sourceTypes;
+        this.partitionTypes = partitionTypes;
+        this.partitionChannels = partitionChannels;
+        this.operatorContext = operatorContext;
+
+        initializeInMemoryGroupedTopNBuilder();
+
+        this.localUserMemoryContext = operatorContext.localUserMemoryContext();
+        this.localRevocableMemoryContext = operatorContext.localRevocableMemoryContext();
+
+        this.unspillMemoryLimit = unspillMemoryLimit;
+        this.producingOutput = false;
+    }
+
+    public Work<?> processPage(Page page)
+    {
+        checkState(hasPreviousSpillCompletedSuccessfully(), "Previous spill hasn't yet finished");
+
+        Work<?> result = inMemoryGroupedTopNBuilder.processPage(page);
+        updateMemoryReservations();
+        return result;
+    }
+
+    private boolean hasPreviousSpillCompletedSuccessfully()
+    {
+        if (spillInProgress.isDone()) {
+            // check for exception from previous spill for early failure
+            checkSpillSucceeded(spillInProgress);
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
+
+    @Override
+    public WorkProcessor<Page> buildResult()
+    {
+        // spill could be in progress.
+        checkSpillSucceeded(spillInProgress);
+
+        producingOutput = true;
+        // Convert revocable memory to user memory as returned Iterator holds on to memory so we no longer can revoke.
+        if (localRevocableMemoryContext.getBytes() > 0) {
+            long currentRevocableBytes = localRevocableMemoryContext.getBytes();
+            localRevocableMemoryContext.setBytes(0);
+            if (!localUserMemoryContext.trySetBytes(localUserMemoryContext.getBytes() + currentRevocableBytes)) {
+                localRevocableMemoryContext.setBytes(currentRevocableBytes);
+                checkSpillSucceeded(spillToDisk());
+            }
+        }
+
+        if (!spiller.isPresent()) {
+            return inMemoryGroupedTopNBuilder.buildResult();
+        }
+
+        // spill could be in progress. Wait till its finished
+        checkSpillSucceeded(spillInProgress);
```

Comment:
Can you provide a scenario when this can happen?

Reply:
You are right. This check is redundant. I moved it to Line 123 but didn't remove it from here. Will update